// @reclaimprotocol/attestor-core/lib/server/socket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttestorServerSocket = void 0;
const message_handler_1 = require("../client/utils/message-handler");
const handlers_1 = require("../server/handlers");
const apm_1 = require("../server/utils/apm");
const generics_1 = require("../server/utils/generics");
const utils_1 = require("../utils");
const socket_base_1 = require("../utils/socket-base");
const util_1 = require("util");
class AttestorServerSocket extends socket_base_1.AttestorSocket {
    constructor(socket, sessionId, bgpListener, logger) {
        // @ts-ignore
        super(socket, {}, logger);
        this.sessionId = sessionId;
        this.bgpListener = bgpListener;
        this.tunnels = {};
        // handle RPC requests
        this.addEventListener('rpc-request', handleRpcRequest.bind(this));
        // forward packets to the appropriate tunnel
        this.addEventListener('tunnel-message', handleTunnelMessage.bind(this));
        // close all tunnels when the connection is terminated
        // since this tunnel can no longer be written to
        this.addEventListener('connection-terminated', () => {
            for (const tunnelId in this.tunnels) {
                const tunnel = this.tunnels[tunnelId];
                tunnel.close(new Error('WS session terminated'));
            }
        });
    }
    getTunnel(tunnelId) {
        const tunnel = this.tunnels[tunnelId];
        if (!tunnel) {
            throw new utils_1.AttestorError('ERROR_NOT_FOUND', `Tunnel "${tunnelId}" not found`);
        }
        return tunnel;
    }
    removeTunnel(tunnelId) {
        delete this.tunnels[tunnelId];
    }
    static async acceptConnection(socket, { req, logger, bgpListener }) {
        // promisify ws.send -- so the sendMessage method correctly
        // awaits the send operation
        const bindSend = socket.send.bind(socket);
        socket.send = (0, util_1.promisify)(bindSend);
        const sessionId = (0, utils_1.generateSessionId)();
        logger = logger.child({ sessionId });
        const client = new AttestorServerSocket(socket, sessionId, bgpListener, logger);
        try {
            const initMsgs = (0, generics_1.getInitialMessagesFromQuery)(req);
            logger.trace({ initMsgs: initMsgs.length }, 'new connection, validating...');
            for (const msg of initMsgs) {
                await message_handler_1.handleMessage.call(client, msg);
            }
            logger.debug('connection accepted');
        }
        catch (err) {
            logger.error({ err }, 'error in new connection');
            if (client.isOpen) {
                client.terminateConnection(err instanceof utils_1.AttestorError
                    ? err
                    : utils_1.AttestorError.badRequest(err.message));
            }
            return;
        }
        return client;
    }
}
exports.AttestorServerSocket = AttestorServerSocket;
async function handleTunnelMessage({ data: { tunnelId, message } }) {
    var _a;
    try {
        const tunnel = this.getTunnel(tunnelId);
        await tunnel.write(message);
    }
    catch (err) {
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.error({
            err,
            tunnelId,
        }, 'error writing to tunnel');
    }
}
async function handleRpcRequest({ data: { data, requestId, respond, type } }) {
    var _a, _b;
    const logger = this.logger.child({
        rpc: type,
        requestId
    });
    const apm = (0, apm_1.getApm)();
    const tx = apm === null || apm === void 0 ? void 0 : apm.startTransaction(type);
    tx === null || tx === void 0 ? void 0 : tx.setLabel('requestId', requestId);
    tx === null || tx === void 0 ? void 0 : tx.setLabel('sessionId', this.sessionId.toString());
    const userId = (_b = (_a = this.metadata.auth) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.id;
    if (userId) {
        tx === null || tx === void 0 ? void 0 : tx.setLabel('authUserId', userId);
    }
    try {
        logger.debug({ data }, 'handling RPC request');
        const handler = handlers_1.HANDLERS[type];
        const res = await handler(data, { client: this, logger, tx });
        await respond(res);
        logger.debug({ res }, 'handled RPC request');
        tx === null || tx === void 0 ? void 0 : tx.setOutcome('success');
    }
    catch (err) {
        logger.error({ err }, 'error in RPC request');
        respond(utils_1.AttestorError.fromError(err));
        tx === null || tx === void 0 ? void 0 : tx.setOutcome('failure');
        apm === null || apm === void 0 ? void 0 : apm.captureError(err, { parent: tx });
    }
    finally {
        tx === null || tx === void 0 ? void 0 : tx.end();
    }
}
