// @reclaimprotocol/attestor-core/lib/client/utils/client-socket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttestorClient = void 0;
const utils_1 = require("ethers/lib/utils");
const config_1 = require("../../config");
const api_1 = require("../../proto/api");
const utils_2 = require("../../utils");
const socket_base_1 = require("../../utils/socket-base");
const ws_1 = require("../../utils/ws");
class AttestorClient extends socket_base_1.AttestorSocket {
    constructor({ url, initMessages = [], signatureType = config_1.DEFAULT_METADATA.signatureType, logger = utils_2.logger, authRequest, makeWebSocket = ws_1.makeWebSocket }) {
        const initRequest = {
            ...config_1.DEFAULT_METADATA,
            signatureType,
            auth: authRequest
        };
        const msg = (0, utils_2.packRpcMessages)({ initRequest }, ...initMessages);
        const initRequestBytes = api_1.RPCMessages.encode(msg).finish();
        const initRequestB64 = utils_1.base64.encode(initRequestBytes);
        url = new URL(url.toString());
        url.searchParams.set('messages', initRequestB64);
        super(makeWebSocket(url), initRequest, logger);
        this.waitForInit = () => {
            if (this.isClosed) {
                throw new utils_2.AttestorError('ERROR_NETWORK_ERROR', 'Client connection already closed');
            }
            return this.waitForInitPromise;
        };
        const initReqId = msg.messages[0].id;
        this.waitForInitPromise = this
            .waitForResponse(initReqId)
            .then(res => {
            logger.info('client initialised');
            this.isInitialised = true;
            this.initResponse = res;
        });
        // swallow the error if anything bad happens, and we've no
        // catch block to handle it
        this.waitForInitPromise
            .catch(() => { });
        this.addEventListener('connection-terminated', ev => (logger.info({ err: ev.data }, 'connection terminated')));
    }
    async rpc(type, request) {
        this.logger.debug({ type }, 'sending rpc request');
        const now = Date.now();
        try {
            const { messages: [{ id }] } = await this.sendMessage({ [(0, utils_2.getRpcRequestType)(type)]: request });
            const rslt = await this.waitForResponse(id);
            return rslt;
        }
        finally {
            const timeTakenMs = Date.now() - now;
            this.logger.debug({ type, timeTakenMs }, 'received rpc response');
        }
    }
    waitForResponse(id) {
        if (this.isClosed) {
            throw new utils_2.AttestorError('ERROR_NETWORK_ERROR', 'Client connection already closed');
        }
        // setup a promise to wait for the response
        return new Promise((resolve, reject) => {
            const handler = (event) => {
                if (event.data.id !== id) {
                    return;
                }
                removeHandlers();
                if ('error' in event.data) {
                    reject(event.data.error);
                    return;
                }
                // @ts-expect-error
                resolve(event.data.data);
            };
            const terminateHandler = (event) => {
                removeHandlers();
                // if the connection was terminated, reject the promise
                // but update the error code to reflect the network error
                if (event.data.code === 'ERROR_NO_ERROR') {
                    reject(new utils_2.AttestorError('ERROR_NETWORK_ERROR', event.data.message, event.data.data));
                    return;
                }
                reject(event.data);
            };
            const removeHandlers = () => {
                this.removeEventListener('rpc-response', handler);
                this.removeEventListener('connection-terminated', terminateHandler);
            };
            this.addEventListener('rpc-response', handler);
            this.addEventListener('connection-terminated', terminateHandler);
        });
    }
}
exports.AttestorClient = AttestorClient;
