// @reclaimprotocol/attestor-core/lib/server/handlers/createTunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTunnel = void 0;
const make_tcp_tunnel_1 = require("../../server/tunnels/make-tcp-tunnel");
const apm_1 = require("../../server/utils/apm");
const dns_1 = require("../../server/utils/dns");
const utils_1 = require("../../utils");
const createTunnel = async ({ id, ...opts }, { tx, logger, client }) => {
    var _a, _b, _c;
    if (client.tunnels[id]) {
        throw utils_1.AttestorError.badRequest(`Tunnel "${id}" already exists`);
    }
    const allowedHosts = (_c = (_b = (_a = client.metadata) === null || _a === void 0 ? void 0 : _a.auth) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.hostWhitelist;
    if ((allowedHosts === null || allowedHosts === void 0 ? void 0 : allowedHosts.length) && !allowedHosts.includes(opts.host)) {
        throw utils_1.AttestorError.badRequest(`Host "${opts.host}" not allowed by auth request`);
    }
    let cancelBgp;
    const apm = (0, apm_1.getApm)();
    const sessionTx = apm === null || apm === void 0 ? void 0 : apm.startTransaction('tunnelConnection', { childOf: tx });
    sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setLabel('tunnelId', id.toString());
    sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setLabel('hostPort', `${opts.host}:${opts.port}`);
    sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setLabel('geoLocation', opts.geoLocation);
    try {
        const tunnel = await (0, make_tcp_tunnel_1.makeTcpTunnel)({
            ...opts,
            logger,
            onMessage(message) {
                if (!client.isOpen) {
                    logger.warn('client is closed, dropping message');
                    return;
                }
                client.sendMessage({
                    tunnelMessage: {
                        tunnelId: id,
                        message
                    }
                });
            },
            onClose(err) {
                cancelBgp === null || cancelBgp === void 0 ? void 0 : cancelBgp();
                if (err) {
                    apm === null || apm === void 0 ? void 0 : apm.captureError(err, { parent: sessionTx });
                    sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setOutcome('failure');
                }
                else {
                    sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setOutcome('success');
                }
                sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.end();
                if (!client.isOpen) {
                    return;
                }
                client.sendMessage({
                    tunnelDisconnectEvent: {
                        tunnelId: id,
                        error: err
                            ? utils_1.AttestorError
                                .fromError(err)
                                .toProto()
                            : undefined
                    }
                })
                    .catch(err => {
                    logger.error({ err }, 'failed to send tunnel disconnect event');
                });
            },
        });
        try {
            await checkForBgp(tunnel);
        }
        catch (err) {
            logger.warn({ err, host: opts.host }, 'failed to start BGP overlap check');
        }
        client.tunnels[id] = tunnel;
        return {};
    }
    catch (err) {
        apm === null || apm === void 0 ? void 0 : apm.captureError(err, { parent: sessionTx });
        sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.setOutcome('failure');
        sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.end();
        cancelBgp === null || cancelBgp === void 0 ? void 0 : cancelBgp();
        throw err;
    }
    async function checkForBgp(tunnel) {
        if (!client.bgpListener) {
            return;
        }
        // listen to all IPs for the host -- in case any of them
        // has a BGP announcement overlap, we'll close the tunnel
        // so the user can retry
        const ips = await (0, dns_1.resolveHostnames)(opts.host);
        cancelBgp = client.bgpListener.onOverlap(ips, (info) => {
            logger.warn({ info, host: opts.host }, 'BGP announcement overlap detected');
            // track how many times we've seen a BGP overlap
            sessionTx === null || sessionTx === void 0 ? void 0 : sessionTx.addLabels({ bgpOverlap: true, ...info });
            tunnel === null || tunnel === void 0 ? void 0 : tunnel.close(new utils_1.AttestorError('ERROR_BGP_ANNOUNCEMENT_OVERLAP', `BGP announcement overlap detected for ${opts.host}`));
        });
        logger.debug({ ips }, 'checking for BGP overlap');
    }
};
exports.createTunnel = createTunnel;
