// @reclaimprotocol/attestor-core/lib/config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BGP_WS_URL = exports.TOPRF_DOMAIN_SEPARATOR = exports.DEFAULT_AUTH_EXPIRY_S = exports.MAX_PAYLOAD_SIZE = exports.MAX_NO_DATA_INTERVAL_MS = exports.PING_INTERVAL_MS = exports.DEFAULT_METADATA = exports.MAX_CLAIM_TIMESTAMP_DIFF_S = exports.DNS_SERVERS = exports.CONNECTION_TIMEOUT_MS = exports.API_SERVER_PORT = exports.DEFAULT_REMOTE_FILE_FETCH_BASE_URL = exports.BROWSER_RPC_PATHNAME = exports.WS_PATHNAME = exports.DEFAULT_HTTPS_PORT = exports.RECLAIM_USER_AGENT = exports.DEFAULT_ZK_CONCURRENCY = exports.MAX_ZK_CHUNKS = void 0;
const api_1 = require("../proto/api");
exports.MAX_ZK_CHUNKS = 200;
exports.DEFAULT_ZK_CONCURRENCY = 10;
exports.RECLAIM_USER_AGENT = 'reclaim/0.0.1';
exports.DEFAULT_HTTPS_PORT = 443;
exports.WS_PATHNAME = '/ws';
exports.BROWSER_RPC_PATHNAME = '/browser-rpc';
exports.DEFAULT_REMOTE_FILE_FETCH_BASE_URL = `${exports.BROWSER_RPC_PATHNAME}/resources`;
exports.API_SERVER_PORT = 8001;
// 10s
exports.CONNECTION_TIMEOUT_MS = 10000;
exports.DNS_SERVERS = [
    '8.8.8.8',
    '8.8.4.4'
];
// 10m
exports.MAX_CLAIM_TIMESTAMP_DIFF_S = 10 * 60;
exports.DEFAULT_METADATA = {
    signatureType: api_1.ServiceSignatureType.SERVICE_SIGNATURE_TYPE_ETH,
    clientVersion: api_1.AttestorVersion.ATTESTOR_VERSION_2_0_0,
    auth: undefined
};
exports.PING_INTERVAL_MS = 10000;
/**
 * Maximum interval in seconds to wait for before assuming
 * the connection is dead
 * @default 30s
 */
exports.MAX_NO_DATA_INTERVAL_MS = 30000;
exports.MAX_PAYLOAD_SIZE = 512 * 1024 * 1024; // 512MB
exports.DEFAULT_AUTH_EXPIRY_S = 15 * 60; // 15m
exports.TOPRF_DOMAIN_SEPARATOR = 'reclaim-toprf';
exports.BGP_WS_URL = 'wss://ris-live.ripe.net/v1/ws/?client=reclaim-hijack-detector';
